### PR TITLE
SCP-3195 - Re implement get total funds through WBE (part 2 frontend)

### PR DESCRIPTION
**Pull request by hrajchert**
This is part 2 of [this PR](https://github.com/input-output-hk/marlowe-cardano/pull/37). It gets both the ada value and the custom tokens (including role tokens) from the WalletBackEnd.

There is an important warning under the `Note [polling updateTotalFunds]` which may cause some bugs while the wallet is syncing. I created two issues to solve that SCP-3211 (add global polling to getTotalFunds) and SCP-3212 (Show wallet sync status in the UI) 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [X] Commit sequence broadly makes sense
    - [X] Key commits have useful messages
    - [X] Relevant tickets are mentioned in commit messages
    - [X] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [X] Self-reviewed the diff
    - [X] Useful pull request description
    - [X] Reviewer requested

### CODE REVIEW

File: marlowe-dashboard-client/src/Page/Dashboard/State.purs [hunk 1]
```diff
@@ -384,6 +395,29 @@ handleAction input@{ currentSlot, tzOffset } (ContractAction followerAppId contr
...
+--       Refactor this code and do a global polling mechanism.
+updateTotalFunds ::
+  forall m slots msg.
+  MonadAff m =>
+  ManageWallet m =>
+  HalogenM State Action slots msg m Unit
+updateTotalFunds = do
+  walletId <- use (_walletDetails <<< _walletInfo <<< _walletId)
+  response <- getWalletTotalFunds walletId
+  for_ response \(GetTotalFunds { assets }) ->
```
Comment (paluh) [inline marlowe-dashboard-client/src/Page/Dashboard/State.purs:417]: We can shorten this to [`Lens.Setter.assign`](https://github.com/purescript-contrib/purescript-profunctor-lenses/blob/v7.0.1/src/Data/Lens/Setter.purs#L96) like above or even just [`.=`](https://github.com/purescript-contrib/purescript-profunctor-lenses/blob/v7.0.1/src/Data/Lens/Setter.purs#L45)

File: marlowe-dashboard-client/src/Capability/Marlowe.purs [hunk 2]
```diff
@@ -89,24 +89,16 @@ instance manageMarloweAppM :: ManageMarlowe AppM where
         -- create the WalletCompanion and MarloweApp for this wallet
         ajaxCompanionAppId <- Contract.activateContract WalletCompanion walletId
         ajaxMarloweAppId <- Contract.activateContract MarloweApp walletId
```
Comment (paluh) [inline marlowe-dashboard-client/src/Capability/Marlowe.purs:91]: Understanding check: Do we drop this because we don't want to event checkt total funds against freshly created wallet?

File: marlowe-dashboard-client/src/Capability/Marlowe.purs [hunk 3]
```diff
@@ -213,15 +205,14 @@ instance manageMarloweAppM :: ManageMarlowe AppM where
             wallet = toFront $ view _cicWallet clientState
           walletContracts <- withExceptT Just $ ExceptT $ Contract.getWalletContractInstances wallet
           walletInfo <- withExceptT Just $ ExceptT $ Wallet.getWalletInfo wallet
```
Comment (paluh) [inline marlowe-dashboard-client/src/Capability/Marlowe.purs:207]: Understanding check: Do we drop this because we want to introduce more general solution for checks (per specific view sync or maybe based on some "app routing" logic etc.)?

File: marlowe-dashboard-client/src/Component/Contacts/State.purs [hunk 4]
```diff
@@ -172,6 +167,24 @@ handleAction (ClipboardAction clipboardAction) = do
   mapAction ClipboardAction $ Clipboard.handleAction clipboardAction
   addToast $ successToast "Copied to clipboard"
```
Comment (paluh) [inline marlowe-dashboard-client/src/Component/Contacts/State.purs:169]: Understanding check: This is just a plain refactoring - replacing an action which was exclusively used by other handler.

File: marlowe-dashboard-client/src/Component/Contacts/State.purs [hunk 5]
```diff
@@ -188,11 +201,24 @@ toWalletIdInput ::
 toWalletIdInput = mapSubmodule _walletIdInput WalletIdInputAction
 
 ------------------------------------------------------------
+-- The cardano Blockchain has Multi-Asset support, each monetary policy is identified
+-- by a policyId which is also known as the CurrencySymbol. The ADA token is a special
+-- case that has the empty string as both the token name and the currency symbol
+-- https://docs.cardano.org/native-tokens/learn
+-- https://github.com/input-output-hk/plutus/blob/1f31e640e8a258185db01fa899da63f9018c0e85/plutus-ledger-api/src/Plutus/V1/Ledger/Ada.hs#L45
```
Comment (paluh) [inline marlowe-dashboard-client/src/Component/Contacts/State.purs:208]: There is already: https://github.com/input-output-hk/purescript-web-common/tree/main/src/Plutus/V1/Ledger

**Review by paluh — COMMENTED** [review]
I've included "self-check" comments with prefix "Understanding check" - these are only checks if I get the idea behind the code.

File: marlowe-dashboard-client/src/Capability/Marlowe.purs (hunk 2 above)
Comment (hrajchert) [reply marlowe-dashboard-client/src/Capability/Marlowe.purs:91]: As you mentioned yesterday, for a real wallet (testnet or even mainnet), it wouldn't make sense

File: marlowe-dashboard-client/src/Capability/Marlowe.purs (hunk 3 above)
Comment (hrajchert) [reply marlowe-dashboard-client/src/Capability/Marlowe.purs:207]: Yeah, as I mentioned in the general description, as part of SCP-3211 I want to have a more general way of getting this information. Probably setting it to the global store and having a heuristic for doing a global poll (either a fixed interval of X seconds or every time there is Y type of event, or similar). Also, we could later think if it makes sense for us to use the chainlink to listen for changes to a particular address.

File: marlowe-dashboard-client/src/Component/Contacts/State.purs (hunk 4 above)
Comment (hrajchert) [reply marlowe-dashboard-client/src/Component/Contacts/State.purs:169]: Yes, it didn't make sense to have it as an Action if it was only used internally

File: marlowe-dashboard-client/src/Component/Contacts/State.purs (hunk 5 above)
Comment (hrajchert) [reply marlowe-dashboard-client/src/Component/Contacts/State.purs:208]: Cool, we can put this next to that

File: marlowe-dashboard-client/src/Page/Dashboard/State.purs (hunk 1 above)
Comment (hrajchert) [reply marlowe-dashboard-client/src/Page/Dashboard/State.purs:417]: Yeah, the only caveat I see is that if we want to modify 2 properties we need to change the "structure" of how we set things. I mean for a single property use assign, for multiple using modify_ + set, why not just use the latter.

Comment (paluh) [reply marlowe-dashboard-client/src/Page/Dashboard/State.purs:417]: I'm not sure I get it... you can always:

```purescript
do
  _lens1 := x
  _lens2 := y
```

Comment (hrajchert) [reply marlowe-dashboard-client/src/Page/Dashboard/State.purs:417]: yes, but I'm almost sure that doing 

```purescript
do
  _lens1 := x 
  _lens2 := y
  _lens3 := z
```

causes 3 renders in Halogen, while doing

```purescript
modify_ $ 
   set _lens1 x 
   <<< set _lens2 y 
   <<< set _lens3 z
```

only triggers 1 render.
